src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs

```diff
@@ -32,7 +32,7 @@ public class InputTagHelper : TagHelper
                 { "Url", "url" },
                 { "EmailAddress", "email" },
                 { "Date", "date" },
-                { "DateTime", "datetime" },
+                { "DateTime", "datetime-local" },
```

Comment:
Thinking about this more, should narrow this change to exactly what @Eilon [said](https://github.com/aspnet/Mvc/issues/4871#issuecomment-235015347):
> change `System.DateTime` to render as `<input type="datetime-local" ... />`

So, remove this change and instead update `TemplateRenderer.GetTypeNames()`. Note `InputTagHelper` and `DefaultEditorTemplates` both use that method. Specifically, change [`TemplateRenderer` lines 208-210](https://github.com/aspnet/Mvc/blob/48240d4b89d5f04a68bb8a5737431405b46dc415/src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/TemplateRenderer.cs#L208-L210) to return `"datetime-local"` for the `DateTime` and `DateTimeOffset` types. That way, `[DataType(...)]` should still work but the defaults will be valid.